### PR TITLE
Re-introduce Push parser

### PR DESCRIPTION
**Pull request by kaidokert**
## Summary by Sourcery

Re-introduce and fully integrate a push-based JSON parser by adding PushParser and its content extractor, extending the core parser to support chunked inputs, and bolting it into existing conformance tests. Enhance buffer and slice management across modules, and provide a wealth of new unit and stress tests along with a demo example to validate and illustrate the new push parsing functionality.

New Features:
- Reintroduce PushParser API with handler-based incremental JSON parsing (PushParser, PushContentBuilder, PushParserHandler, PushParseError)
- Add streaming JSON conformance tests for PushParser and StreamParser in json_checker_tests
- Provide a push_parser_demo example demonstrating chunked data feeding and SAX-style event handling

Enhancements:
- Extend ParserCore with a handles_chunked_input flag and separate constructors for chunked vs non-chunked parsing
- Add shared.as_bytes method and StreamBuffer.truncate_unescaped_by for improved buffer management
- Strengthen DataSource get_borrowed_slice bounds checking to support chunked input

Documentation:
- Include a push_parser_demo.rs example in the examples directory

Tests:
- Add extensive PushParser unit tests covering conformance, escape sequences, Unicode, copy-on-escape optimization, and error cases
- Introduce PushParser stress tests for various buffer sizes, chunk patterns, and edge conditions
- Expand json_checker_tests to include pass/fail cases for PushParser and StreamParser via a test macro
- Add StreamParser InputBufferFull test scenarios for oversized inputs

## Summary by CodeRabbit

* **New Features**
  * Introduced a streaming, event-driven (SAX-style) push parser for JSON, supporting incremental input and robust escape sequence handling.
  * Added public APIs for push-based parsing, including new handler traits and error types.
  * Provided a demonstration example for using the push parser with streaming input.

* **Bug Fixes**
  * Improved handling of buffer boundaries and escape sequences to prevent errors with chunked input.

* **Tests**
  * Added extensive tests for the push parser, covering escape handling, chunked input, buffer overflows, conformance, known deviations, and stress scenarios.

* **Documentation**
  * Included example usage and enhanced test coverage to illustrate new parser capabilities.

### CODE REVIEW

File: picojson/src/push_content_builder.rs [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+        // Get the slice of partial content from the current chunk
+        if content_end > content_start {
+            let partial_slice = self.get_borrowed_slice(content_start, content_end)?;
+
+            // Copy bytes to a fixed-size buffer to avoid borrowing conflicts
+            let mut temp_buffer = [0u8; 1024]; // Should handle most token sizes
+            let byte_count = partial_slice.len();
+            if byte_count > temp_buffer.len() {
+                return Err(ParseError::InputBufferFull);
+            }
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_content_builder.rs]: Removed the buffer

File: picojson/src/push_content_builder.rs [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+
+                        // Remove the first hex digit - use small bounded buffer
+                        let mut temp_content = [0u8; 64]; // Small bounded buffer for reasonable string prefixes
+                        let content_len_without_hex = current_content.len() - 1;
+
+                        if content_len_without_hex > temp_content.len() {
+                            // String too long - this shouldn't happen for typical use cases
+                            // Just clear everything and continue
+                            self.stream_buffer.clear_unescaped();
+                        } else if content_len_without_hex > 0 {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_content_builder.rs]: Removed this buffer too

File: picojson/src/push_content_builder.rs [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+    }
+
+    fn has_unescaped_content(&self) -> bool {
+        self.using_unescaped_buffer
+    }
+}
+
+impl PushContentExtractor<'_, '_> {
+    /// Copy partial content from current chunk to scratch buffer when chunk boundary reached
+    pub fn copy_partial_content_to_scratch(&mut self) -> Result<(), ParseError> {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply picojson/src/push_content_builder.rs:493]: Okay what about this comment though - close / resolve or not ?
